From ff09c1131824e5d4757adf69c0c4761fa6040ee6 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Tue, 16 Aug 2016 14:15:20 +0300
Subject: [PATCH 0412/1345] fix: mvpp2x: limit TX coalescing time

commit  022b07d75395b4f7d3cee733dc9706881d48e539 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- limit TX coalescing time with 0xfffff0 core clock cycles

Change-Id: Ica641f032aec8ca9b9d290bc22500748fb8f8c4b
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31970
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c   |    2 ++
 .../net/ethernet/marvell/mvpp2x/mv_pp2x_hw_type.h  |    3 ++-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c
index 281af1c..c79ef38 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.c
@@ -3511,6 +3511,8 @@ void mv_pp2x_tx_done_time_coal_set(struct mv_pp2x_port *port, u32 usec)
 	u32 val;
 
 	val = (port->priv->hw.tclk / USEC_PER_SEC) * usec;
+	if (val > MVPP22_MAX_ISR_TX_THRESHOLD)
+		val = MVPP22_MAX_ISR_TX_THRESHOLD;
 	mv_pp2x_write(&port->priv->hw,
 		      MVPP22_ISR_TX_THRESHOLD_REG(port->id), val);
 }
diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw_type.h b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw_type.h
index 528eab6..e322711 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw_type.h
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw_type.h
@@ -846,7 +846,8 @@
 
 /* Interrupt Cause and Mask registers */
 #define MVPP22_ISR_TX_THRESHOLD_REG(port)	(0x5140 + 4 * (port))
-#define MVPP22_ISR_TX_THRESHOLD_MASK		0xfffff0
+#define MVPP22_MAX_ISR_TX_THRESHOLD		0xfffff0
+#define MVPP22_ISR_TX_THRESHOLD_MASK		MVPP22_MAX_ISR_TX_THRESHOLD
 
 #define MVPP2_ISR_RX_THRESHOLD_REG(rxq)		(0x5200 + 4 * (rxq))
 #define MVPP2_ISR_RX_THRESHOLD_MASK		0xfffff0
-- 
1.7.9.5

